lib/utils/aws/organizations/matchingaccounts.go

```diff
@@ -0,0 +1,288 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package organizations
+
+import (
+	"context"
+	"slices"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/organizations"
+	organizationstypes "github.com/aws/aws-sdk-go-v2/service/organizations/types"
+	"github.com/gravitational/trace"
+)
+
+// RequiredAPIs lists the AWS Organizations APIs required by MatchingAccounts.
+func RequiredAPIs() []string {
+	return []string{
+		"organizations:ListChildren",
+		"organizations:ListRoots",
+		"organizations:ListAccounts",
+		"organizations:ListAccountsForParent",
+	}
+}
+
+// OrganizationsClient is a minimal subset of AWS Organizations API used by MatchingAccounts.
+type OrganizationsClient interface {
+	organizations.ListChildrenAPIClient
+	organizations.ListRootsAPIClient
+	organizations.ListAccountsAPIClient
+	organizations.ListAccountsForParentAPIClient
+}
+
+type awsOrgItem struct {
+	id                  string
+	organizationalUnits []*awsOrgItem
+	accounts            []string
+}
+
+// MatchingAccountsFilter defines the filter to apply when retrieving matching accounts from an AWS Organization.
+type MatchingAccountsFilter struct {
```

Comment:
Some other types similar to `MatchingAccountsFilter` have a very useful `match()` method that can simplify the usage of the type; as it is, the type mostly keeps the data but the validation logic is fuzzy.

Reply:
It allows us to ignore part of the tree which we don't care about because it was excluded.

The alternative would be to always fetch the entire tree and only filter afterwards which would lead to more API calls.